Cloud Native Security Whitepaper submitted (#452)
cncf/sig-security: security-whitepaper/cloud-native-security-whitepaper.md

Shared with CNCF Community

# **Cloud Native Security Whitepaper**

|
**Work-Group:** SIG-Security **Short self link to working doc:** https://tinyurl.com/CNCF-CNSWP | **Status** : WIP | In Review | **Approved** | Obsolete

**Created** : 2020-FEB-01 **Reviewed** : 2020-OCT-27  **Published** : 2020-NOV-18 **Release Version:** 1.0 **Approvers** : [x] @lizrice [x] @justincormack  |

> | **Contributors:** [aradhna.chetal@gmail.com](mailto:aradhna.chetal@gmail.com), [themoxiefoxatwork@gmail.com](mailto:themoxiefoxatwork@gmail.com), [jj@tetrate.io](mailto:jj@tetrate.io), gadi@alcide.io @lumjjb, @trishankatdatadog, [@vvenkatara@paloaltonetworks.com](mailto:vvenkatara@paloaltonetworks.com), @pushkarj-v, @whaber, @sublimino, @rowan-baker, [chase.pettet@gmail.com](mailto:chase.pettet@gmail.com), [harsingh@us.ibm.com](mailto:harsingh@us.ibm.com), jeff.lombardo@gmail.com

Reviewer: @idvoretskyi : Had a recent change in my Github ID/username and the one here does not exist. Can you please replace it with `@pushkarj-v` with `@pushkarj`. Thanks and sorry for the bother.

Reviewer 2: Can we add links to the github IDs as well?

### Deploy

![Figure 4](RackMultipart20201111_figure4.png)

_Figure 4_

The "Deploy" phase is responsible for incorporating a sequence of 'pre-flight' checks in order to ensure that the applications that are going to be deployed in the runtime environment conform and comply with organization wide security and compliance policies.

#### Pre-Flight Deployment Checks

Prior to deployment, organizations should verify the existence, applicability, and current state of:

- Image signature and integrity
- Image runtime policies (e.g absence of malware or critical vulnerabilities)
- Container runtime policies (e.g absence of excessive privileges)
- Host vulnerability and compliance controls
- Workload, application, and network security policies

#### Observability & Metrics

> Instituting observability and metrics into cloud native architectures delivers security insights so appropriate stakeholders can resolve and mitigate anomalies appearing in reporting; tools in this area can help collect and visualize this information. Through the use of behavioral and heuristic analysis teams can detect and escalate outliers, suspicious events, and unexplained calls to appropriate stakeholders. Artificial intelligence (AI), [machine learning (ML)](#kix.tlhq6ssgript), or statistical modeling are all mechanisms that are encouraged to assist in behavioral and heuristic analysis development.

Reviewer: Is the referenced link for ML intentional ?

Author: @PushkarJ it's not - will be fixed.

#### Service Mesh

A service mesh provides connectivity between the services that adds additional capabilities like traffic control, service discovery, load balancing, resilience, observability, security, and so on. A service mesh allows microservices to offload these capabilities from application-level libraries and allows developers to focus on differentiating business logic. In order to effectively ensure secure communications between services in cloud native environments, organizations should implement a service mesh to [eliminate implicit trust](#kix.kyffcmw3zhk)within their pods and across workloads, achieved through data-in-motion encryption. Utilization of a service mesh also resolves identity issues where traditional layer 3 and layer 4 identities, IP addresses, no longer cleanly map to workloads. Service mesh provides not only network level isolation and security but also network-level resiliency capabilities such as retry, timeout, and implementing various circuit-breaker capabilities. Streaming platforms can benefit from a service mesh for added security by using workload level authorization to set access rules for topics or brokers.

> It is important to note that implementation of a [service mesh](#_8999rr470gwb) can help reduce the attack surface of a cloud native deployment, and provide a key framework for building [zero trust](#kix.kyffcmw3zhk)application networks.

Reviewer: Missing space between `zero trust` and `application`

- Chase Pettet - [Mirantis](https://www.mirantis.com/) (Chase.Pettet@mirantis.com)
- Emily Fox - [US National Security Agency (NSA)](https://www.nsa.gov/)
- Gadi Naor - [Alcide](http://www.alcide.io/)
- Harmeet Singh - [IBM](https://www.ibm.com/)
- Jeff Lombardo - Independent
- Jeyappragash JJ - [Tetrate IO](https://www.tetrate.io/)
- Pushkar Joglekar - [Visa](https://visa.com/)
- Rowan Baker - [ControlPlane](https://control-plane.io/)
- Andrew Martin - [ControlPlane](https://control-plane.io/)
- Trishank Karthik Kuppusamy - [Datadog](https://www.datadoghq.com/)
- Vinay Venkataraghavan -[Prisma Cloud (Palo Alto Networks)](https://www.paloaltonetworks.com/prisma/cloud)
- Wayne Haber - [GitLab](https://about.gitlab.com/)
- Mark Bower
- Alex Chircop - StorageOS

Reviewers:

- @justincappos
- @lumjjb
- @whaber
- @craigbox
- @anvega
- @magnologan
- Alok Raj - [XenonStack](https://www.xenonstack.com/) ([alok@xenonstack.com](mailto:alok@xenonstack.com))
- @nyrahul
- @ranio1

[1](#sdfootnote1anc) Another model to consider is Cloud, Clusters, Containers, and Code: [https://kubernetes.io/docs/concepts/security/overview/](https://kubernetes.io/docs/concepts/security/overview/)

> [2](#sdfootnote2anc)[Shifting security left](https://www.devsecops.org/blog/2016/5/20/-security) often leaves organizations to lapse operational security monitoring. It is important that security exists in all parts of the lifecycle and organizations continually evaluate other aspects of their business and technology processes where they may reach beyond modern security paradigms to embrace security as a culture and habit.

Reviewer: Needs a space like this:
`[2](#sdfootnote2anc) [Shifting`

> [3] Example - [MITRE ATT&CK Framework for Kubernetes](https://www.darkreading.com/threat-intelligence/microsofts-kubernetes-threat-matrix-heres-whats-missing/a/d-id/1339106)

Reviewer: possible numbering issue because of duplicate 3

Author: The same footnote numbering is in the original document as well. @TheFoxAtWork can you please check?

[3](#sdfootnote3anc) Human capital is a vital asset necessary to the success of any organization, the corresponding intellectual property and relational capital brought as a result is equally in need of protection.

[4](#sdfootnote4anc) https://blog.aquasec.com/malicious-container-image-docker-container-host

[5](#sdfootnote5anc) According to Applied Software Measurement, Capers Jones, 1996 and adjusting for inflation - 85% of defects are introduced during coding with a cost of $41 to fix compared to a post release fix cost of $26,542.

> [6](#sdfootnote6anc) CIsecurity.org maintains a listing of benchmarks for hardening

Reviewer: This should be cisecurity.org without upper case as it does _not_ represent CI (Continuous Integration).

Contributors:

- Aradhna Chetal - [TIAA](http://tiaa.org/)
- Brandon Lum - [IBM](https://www.ibm.com/)
- Chase Pettet - [Mirantis](https://www.mirantis.com/) (Chase.Pettet@mirantis.com)
- Emily Fox - [US National Security Agency (NSA)](https://www.nsa.gov/)
- Gadi Naor - [Alcide](http://www.alcide.io/)
- Harmeet Singh - [IBM](https://www.ibm.com/)
- Jeff Lombardo - Independent
- Jeyappragash JJ - [Tetrate IO](https://www.tetrate.io/)
- Pushkar Joglekar - [Visa](https://visa.com/)
- Rowan Baker - [ControlPlane](https://control-plane.io/)
- Andrew Martin - [ControlPlane](https://control-plane.io/)
- Trishank Karthik Kuppusamy - [Datadog](https://www.datadoghq.com/)
- Vinay Venkataraghavan -[Prisma Cloud (Palo Alto Networks)](https://www.paloaltonetworks.com/prisma/cloud)
- Wayne Haber - [GitLab](https://about.gitlab.com/)
- Mark Bower
- Alex Chircop - StorageOS

Reviewers:

- @justincappos
- @lumjjb
- @whaber
- @craigbox
- @anvega
- @magnologan
- Alok Raj - [XenonStack](https://www.xenonstack.com/) ([alok@xenonstack.com](mailto:alok@xenonstack.com))
- @nyrahul
- @ranio1

> 

Reviewer 3: ```suggestion
- @lizrice 
- @justincormack 

```

Within the cloud native runtime environment, the microservice architecture for applications has been adopted by industries and organizations worldwide. Applications are often composed of several independent and single purpose microservices which communicate with each other via service layer abstractions which the container orchestration layer makes possible. Best practices to secure this interrelated component architecture involves ensuring that only sanctioned processes operate within a container namespace, prevention and notification of unauthorized resource access, and network traffic monitoring to detect hostile tooling activity. Service Mesh is another common abstraction that provides consolidated and complementary functionality for orchestrated services without imposing changes on the workload software itself (e.g. logging of API traffic, transport encryption, observability tagging, authentication, and authorization).

## Recommendations

> Cloud native security seeks to ensure the same conditions of diligence, integrity, trust, and threat prevention as traditional security models while integrating modern concepts of ephemerality, distribution, and immutability. In these rapidly changing environments, prone to fail-forward for iteration, automation inline with the development pipeline is required for secure outcomes. Organizations must earnestly analyze and apply these core security concepts to alleviate delay in applying hardening and environmental controls, and need to hold engaged third parties to the same standard while balancing perpetual education and training relevant to the cloud capabilities and security proponents for their own workforce.With additional layers of complexity and a broad mesh of components to care for, protection from unauthorized access must be accomplished by integrating security throughout the lifecycle and into the runtime environment. It is highly recommended organizations evaluate the security defense stack against the relevant attack frameworks [^3] to achieve clarity about which threats a defense stack covers. Additionally, organizations need to adopt approaches and methodologies that shift security left [^2], amplify DevOps, and reach beyond to the next technology horizon so continued, proper checks of all components before, within, and after the pipeline are verified with any innovation brought into the lifecycle.

Reviewer 3: ```suggestion
Cloud native security seeks to ensure the same conditions of diligence, integrity, trust, and threat prevention as traditional security models while integrating modern concepts of ephemerality, distribution, and immutability. In these rapidly changing environments, prone to fail-forward for iteration, automation inline with the development pipeline is required for secure outcomes. Organizations must earnestly analyze and apply these core security concepts to alleviate delay in applying hardening and environmental controls, and need to hold engaged third parties to the same standard while balancing perpetual education and training relevant to the cloud capabilities and security proponents for their own workforce.With additional layers of complexity and a broad mesh of components to care for, protection from unauthorized access must be accomplished by integrating security throughout the lifecycle and into the runtime environment. It is highly recommended organizations evaluate the security defense stack against the relevant attack frameworks [^2] to achieve clarity about which threats a defense stack covers. Additionally, organizations need to adopt approaches and methodologies that shift security left [^3], amplify DevOps, and reach beyond to the next technology horizon so continued, proper checks of all components before, within, and after the pipeline are verified with any innovation brought into the lifecycle.
```

##### Static Analysis and Security Testing

Static analysis of IaC, application manifests, and software code may cover linting, identifying misconfigurations and vulnerability scanning. IaC code should be subject to the same pipeline policy controls as do application workloads.

IaC is gaining popularity and its implementation is rapidly increasing among organizations to deploy cloud and container infrastructure. Consequently, insecure configurations in these templates can result in exposing attack vectors.

These templates should be scanned for insecure configurations and other security controls using automated tools before deploying the application and infrastructure artifacts. Key misconfigurations to keep an eye out for include:

- Vulnerabilities contained within images specified in the application manifests
- Configuration settings, such as containers that can escalate privileges.
- Identification of the security contexts and system calls, which can compromise a system.
- Resource limit settings

##### Dynamic Analysis

> Dynamic analysis of deployed infrastructure may include detecting RBAC and IAM configuration drift, validating the expected network attack surface, and ensuring that a SOC can detect unusual behavior in dedicated test environments to configure alerting for production. Dynamic analysis is considered to be a part of testing however, it is expected to occur in a non-production runtime environment.

Reviewer 2: ```suggestion
Dynamic analysis of deployed infrastructure may include detecting Role-based Access Control (RBAC) and IAM configuration drift, validating the expected network attack surface, and ensuring that a SOC can detect unusual behavior in dedicated test environments to configure alerting for production. Dynamic analysis is considered to be a part of testing however, it is expected to occur in a non-production runtime environment.
```
RBAC first use

##### Secrets Encryption

It is possible to manage secrets in a container orchestration or deployment environment through use of an external secrets manager or natively using the orchestrator's secrets. When using a native secret store, it is crucial to be aware that several different protection methods are available:

- Encryption with an external Key Management Store (KMS)
  - Leveraging a KMS is a secure way to protect secrets in the orchestrator secret store where key encryption in an external KMS encrypts the Data Encryption Key (DEK) that encrypts the secrets stored at rest in etcd. This method does have an option to cache DEKs in memory to reduce the dependency on the availability of the external KMS and faster decryption of secrets during pod creation time.
- Encryption with the key in a config file
  - This methodology encrypts the secrets stored in the orchestrator, but storage of the encryption key is available in a configuration file on manager nodes.
- No encryption
  - For example, with some orchestrators, secrets are base64 encoded and stored in clear-text in the key-value store by default

> Using an external secrets manager can limit the risks of using unencrypted secrets and ease the complexity of key management. Most of the time those tools are provided as controllers or operators for Kubernetes that can inject secrets at runtime and handle their rotations transparently.

Reviewer 2: ```suggestion
Using an external secrets manager can limit the risks of using unencrypted secrets and ease the complexity of key management. Most of the time those tools are provided as controllers or operators that can inject secrets at runtime and handle their rotations transparently.
```

Storage systems contain a data access interface that defines how applications or workloads store or consume data that is persisted by the storage system or service. This interface can be protected by access controls, authentication, authorization, and potentially encryption in transit.

Storage systems also contain a control plane management interface which is typically an API protected by authentication and TLS, although finer grained access may be available. In general the control interface is only accessed via a service account by an orchestrator or service broker.

##### Storage Stack

Any storage solution is composed of multiple layers of functionality that define how data is stored, retrieved, protected and interacts with an application, orchestrator and/or operating system. Each of these layers has the potential to influence and impact the security of the storage system. A common example may be a filesystem that persists files or blocks to an object store. It is equally important to protect every layer in the topology, and not just the top layer where data is accessed.

###### Orchestration

> Most orchestrated systems will implement a variety of abstraction and virtualization layers that may include filesystems (such as bind mounts), volume managers, and the application of permissions at a user or group level based on orchestrator policies. As with many components of containerization and microservice architectures, protecting volumes and storage will always rely on the protections in place from other capabilities. If a user is able to escalate their privileges within the orchestrator or container runtime to root they can wreak havoc within the environment. The implementation of [zero trust](https://docs.google.com/document/d/1aAkuhZMJ6-Xq6xP0gh1wxIM8OhUETqEFLhoRdxYHN5c/edit#bookmark=kix.kyffcmw3zhk), [least privilege](https://docs.google.com/document/d/1aAkuhZMJ6-Xq6xP0gh1wxIM8OhUETqEFLhoRdxYHN5c/edit#bookmark=id.bl312cp5xyv1), and [access control](https://docs.google.com/document/d/1aAkuhZMJ6-Xq6xP0gh1wxIM8OhUETqEFLhoRdxYHN5c/edit#bookmark=kix.3f703zlu80we) and enforcement are linchpins in successfully securing storage in cloud native architectures.

Reviewer 2: ```suggestion
Most orchestrated systems will implement a variety of abstraction and virtualization layers that may include filesystems (such as bind mounts), volume managers, and the application of permissions at a user or group level based on orchestrator policies. As with many components of containerization and microservice architectures, protecting volumes and storage will always rely on the protections in place from other capabilities. If a user is able to escalate their privileges within the orchestrator or container runtime to root they can wreak havoc within the environment. The implementation of [zero trust](#zero-trust-architecture), [least privilege](#least-privilege), and [access control](#access) and enforcement are linchpins in successfully securing storage in cloud native architectures.
```

- Chase Pettet - [Mirantis](https://www.mirantis.com/) (Chase.Pettet@mirantis.com)
- Emily Fox - [US National Security Agency (NSA)](https://www.nsa.gov/)
- Gadi Naor - [Alcide](http://www.alcide.io/)
- Harmeet Singh - [IBM](https://www.ibm.com/)
- Jeff Lombardo - Independent
- Jeyappragash JJ - [Tetrate IO](https://www.tetrate.io/)
- Pushkar Joglekar - [Visa](https://visa.com/)
- Rowan Baker - [ControlPlane](https://control-plane.io/)
- Andrew Martin - [ControlPlane](https://control-plane.io/)
- Trishank Karthik Kuppusamy - [Datadog](https://www.datadoghq.com/)
- Vinay Venkataraghavan -[Prisma Cloud (Palo Alto Networks)](https://www.paloaltonetworks.com/prisma/cloud)
- Wayne Haber - [GitLab](https://about.gitlab.com/)
- Mark Bower
- Alex Chircop - StorageOS

Reviewers:

- @justincappos
- @lumjjb
- @whaber
- @craigbox
- @anvega
- @magnologan
- Alok Raj - [XenonStack](https://www.xenonstack.com/) ([alok@xenonstack.com](mailto:alok@xenonstack.com))
- @nyrahul
- @ranio1

[1^]: Another model to consider is Cloud, Clusters, Containers, and Code: [https://kubernetes.io/docs/concepts/security/overview/](https://kubernetes.io/docs/concepts/security/overview/)

> [2^]: [Shifting security left](https://www.devsecops.org/blog/2016/5/20/-security) often leaves organizations to lapse operational security monitoring. It is important that security exists in all parts of the lifecycle and organizations continually evaluate other aspects of their business and technology processes where they may reach beyond modern security paradigms to embrace security as a culture and habit.
> 
> [3^]: Example - [MITRE ATT&CK Framework for Kubernetes](https://www.darkreading.com/threat-intelligence/microsofts-kubernetes-threat-matrix-heres-whats-missing/a/d-id/1339106)

Reviewer 3: ```suggestion
[2^]: Example - [MITRE ATT&CK Framework for Kubernetes](https://www.darkreading.com/threat-intelligence/microsofts-kubernetes-threat-matrix-heres-whats-missing/a/d-id/1339106)

[3^]: [Shifting security left](https://www.devsecops.org/blog/2016/5/20/-security) often leaves organizations to lapse operational security monitoring. It is important that security exists in all parts of the lifecycle and organizations continually evaluate other aspects of their business and technology processes where they may reach beyond modern security paradigms to embrace security as a culture and habit.
```
